walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -27,10 +27,33 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        List<TimeRange> mandatoryAvailableRanges = new ArrayList<>();
+        List<TimeRange> optionalAttendeesConsideredRanges = new ArrayList<>();
+
+        /* try to remove the timeRanges optional attendees can't make it to. If there's nothing left, ignore the optional attendees*/
+        mandatoryAvailableRanges = availableRanges(events, request, false);
+        optionalAttendeesConsideredRanges = availableRanges(events, request, true);
+
+        if (request.getAttendees().isEmpty() ) { // only optional Attendees
+            return optionalAttendeesConsideredRanges;
+        } else if (optionalAttendeesConsideredRanges.isEmpty() && !request.getOptionalAttendees().isEmpty()) { // consider optional Attendees but considering optionalAttendees create no options
+            return mandatoryAvailableRanges;
+        } else {
+            return optionalAttendeesConsideredRanges; // last case: there are optionalAttendees to consider and we can consider optionalAttendees + still have avaliabilities
+        }
+  }
+
+  public List<TimeRange> availableRanges(Collection<Event> events, MeetingRequest request, boolean considerOptionalAttendees) {
         List<TimeRange> availableRanges = new ArrayList<>();
         availableRanges.add(TimeRange.WHOLE_DAY);
 
-        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+        Collection<TimeRange> bookedRanges;
+
+        if (considerOptionalAttendees) {
+            bookedRanges = combinedRanges(events, request, true);
+        } else {
+            bookedRanges = combinedRanges(events, request, false);
+        }
```

Comment:
You can simplify this into `Collection<TimeRange> bookedRanges = combinedRanges(events, request, considerOptionalAttendees);`

Reply:
Thanks